### PR TITLE
[Stabilization, Crash] Fix crash when NavBar navigating to symbols without source locations

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes internal bug 187905

In a Visual Basic WinForms application, the Application.Designer.vb file
contains one Partial of the MyApplication class. Using the NavBar in
this class to attempt navigation to “Main” crashes VS. This crash will
happen any time a non-source symbol (that’s also not a “code generation”
entry like “New” and “Finalize”) finds its way into the NavBar, but this
is the only example I currently have.

This has been broken since Visual Studio 2015 RTM, but previous versions
of Visual Studio showed a "The definition of the object is hidden."
dialog in this case.

**Triage**: This fixes a crashing bug that we have Watson data for. It has been crashing for all of Dev14, but used to work in Dev12.

### CODE REVIEW

**Comment by dpoeschl** [comment]
Please review, @dotnet/roslyn-ide 

A couple remaining items:
1. Working on adding a test.
2. Dev12 and prior showed a dialog. Should we do this too?
  image

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb [hunk 1]
```diff
@@ -488,6 +482,19 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
             Return New VirtualTreePoint(location.SourceTree, location.SourceTree.GetText(cancellationToken), location.SourceSpan.Start)
         End Function
 
+        Private Shared Function GetSourceNavigationLocation(document As Document, symbol As ISymbol, cancellationToken As CancellationToken) As Location
+            Dim sourceLocations = symbol.Locations.Where(Function(l) l.IsInSource)
```
Comment (dpoeschl) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb]: The only actual change is to only consider `IsInSource` locations. Extracted this bit to a method to make sure the setting of the "location" and "sourceLocations" locals didn't leak out to the rest of the method.

File: src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb [hunk 2]
```diff
@@ -464,13 +464,7 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
...
             End If
 
-            ' First figure out the location that we want to grab considering partial types
-            Dim location = symbol.Locations.FirstOrDefault(Function(l) l.SourceTree.Equals(document.GetSyntaxTreeAsync(cancellationToken).WaitAndGetResult(cancellationToken)))
-
-            If location Is Nothing Then
-                location = symbol.Locations.FirstOrDefault
-            End If
-
+            Dim location As Location = GetSourceNavigationLocation(document, symbol, cancellationToken)
```
Comment (Pilchie) [inline src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicNavigationBarItemService.vb]: Nit: You could let the type be inferred and leave out the As clause.

**Comment by Pilchie** [comment]
:+1:

**Comment by CyrusNajmabadi** [comment]
:+1: 

Though i think a better fix would be to not put symbols that don't have source locations int he nav bar to begin with.

**Comment by dpoeschl** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/5111/
// Retest reason:

**Comment by dpoeschl** [comment]
@MattGertz @Pilchie Can we consider this for 1.2?

**Comment by MattGertz** [comment]
@dpoeschl What's the hit count, and can you point me at the VSO Watson bug offline?  We can consider for escrow if hit count is high; we'll probably have at least one payload there. @davkean for FYI.

**Comment by MattGertz** [comment]
Approved; please work with @davkean and @natidea on Escrow particulars (how/when/where to merge, what info is needed in the DevDiv bug, etc.)

**Comment by dpoeschl** [comment]
retest vsi please

**Comment by Pilchie** [comment]
retest eta please

**Pull request merged**